Постановка целей

sta1er opened the pull request:
Во время задачи возникло много не понимай и возможно тут написан полный бред. Я старался максимально использовать данные нам методы в интерфейсах GoalRepository(даже дописал два), но не понимаю на сколько это правильно. Так же как и не понимаю зачем мы используем create() из GoalRepository, если есть save.

Commit:
Создать новую цель and Обновить цель

Commit:
Удалить цель(deleteGoal)

Commit:
Получить все подзадачи цели по фильтру

Commit:
Получить список целей по фильтрам

Commit:
Постановка целей
Во время задачи возникло много не понимай и возможно тут написан полный бред. Я старался максимально использовать данные нам методы в интерфейсах GoalRepository(даже дописал два), но не понимаю на сколько это правильно. Так же как и не понимаю зачем мы используем create() из GoalRepository, если есть save.

Commit:
Постановка целей. Исправденный туст(deleteGoal_Success)

Commit:
Постановка целей. Переработаны методы в GoalController и GoalService. Добавлен отдельный  класс для валидации. Полностью переработана фильтрация, сделаны отдельные интерфейсы и классы для фильтрации.

vang31 commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,90 @@
+import java.util.stream.Stream;
+
+@Service
+@Transactional
```
а зачем здесь эта аннотация? вроде в лекциях еще такого не показывали)

sta1er replied:
Ну я знаю что аннотация Transactional помогает при каких-либо проблемах с бд при работе с ней в пределах метода. Условно если мы что-то поменяли -> потом произошла ошибка(между обращениями к бд) -> то эта аннотация откатит изменения. Но я понял почему конкретно здесь ее не желательно вешать на класс, а лучше раскидать только по некоторым методам.

vang31 commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,48 @@
+        this.goalService = goalService;
+    }
+
+    @PostMapping("/create")
```
енд поинт не должен называться так) хоть все они и круд, но существуют манифесты по нэймингу

InnMish replied:
@vang31 а расскажи, какие?)

vang31 commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/goals")
```
/api нужно вынести в yaml/ что бы у каждого контроллера такое не писать. 
Спринг будет сам из ямла подбирать  Spring mvc-context,(посмотри как это делается

InnMish replied:
@vang31 а расскажи, как? хехехехе 
я подскажу: spring servlet mvc path, кажется

Commit:
Правки:
1. Перераспределил аннотацию Transactional по отдельным методам
2. Изменил названия эндпойнтов
3. вынес /api в yaml/

vang31 commented on src/test/java/school/faang/user_service/service/goal/FilterTest.java:
```diff
@@ -0,0 +1,221 @@
+
+        Goal goal3 = new Goal();
+        goal3.setStatus(GoalStatus.ACTIVE);
+        goal3.setId(3L);
```
можно вынести в отдельный метод.. который будет просто создавать Goal. или их список)

vang31 commented on src/test/java/school/faang/user_service/service/goal/FilterTest.java:
```diff
@@ -0,0 +1,221 @@
+        goal3.setStatus(GoalStatus.ACTIVE);
+        goal3.setId(3L);
+
+        List<Goal> list = List.of(goal1, goal2, goal3);
```
как раз их список. выносите в отдельный приватный метод, fillGoal() возвращает List<Goal> 
так будет чище тест)

Commit:
оптимизировал тесты

vang31 commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,89 @@
+    private final GoalValidator goalValidator;
+    private final GoalMapper goalMapper;
+    private final List<GoalFilter> goalFilters;
+    @Autowired
```
отступ нужно добавить

vang31 commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,89 @@
+        this.goalMapper = goalMapper;
+        this.goalFilters = goalFilters;
+    }
+    @Transactional
```
отступ

vang31 commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,89 @@
+        return goalMapper.toDto(saveGoal);
+    }
+    @Transactional
+    public GoalDto updateGoal(Long goalId, GoalDto goalDto) {
```
отступ между 42 и 43ей строкой

vang31 commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,89 @@
+
+        return goalMapper.toDto(savedGoal);
+    }
+    @Transactional
```
+

vang31 commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/goals")
+public class GoalController {
+    private final GoalService goalService;
```
Запускал уже запросы в PostMan?

sta1er replied:
Еще нет, я видел что мы на следующей неделе проходим его, но если мне хватит времени обязательно попробую, правда уже наверное на таске с вакансиями

vang31 commented on src/main/java/school/faang/user_service/repository/goal/GoalRepository.java:
```diff
@@ -49,4 +49,15 @@ WITH RECURSIVE subtasks AS (
+            """)
+    void removeSkillsFromGoal(long goalId);
+
+    @Query(nativeQuery = true, value = """
+            INSERT INTO goal_skill (goal_id, skill_id, created_at, updated_at) 
+            VALUES (?1, ?2, NOW(), NOW())
+            """)
+    void addSkillToGoal(long goalId, long skillId);
```
точно ли нужно делать инсёрт? можно использовать просто save.

sta1er replied:
Вот я долго над этим думал, просто для чего тогда добавили в repository этот метод

vang31 commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,89 @@
+        goalRepository.removeSkillsFromGoal(goalId);
+        goalDto.getSkillIds().forEach(skillId -> goalRepository.addSkillToGoal(savedGoal.getId(), skillId));
+
+        if (goalDto.getStatus().equals("completed")) {
+            List<User> users = goalRepository.findUsersByGoalId(goalId);
+            users.forEach(user -> goalDto.getSkillIds().forEach(skillId -> skillService.assignSkillToUser(skillId, user.getId())));
```
есть ли смысл создавать отдельную коллекцию ? 
можно сразу goalRepository.findUsersByGoalId(goalId).forEach(user -> goalDto.getSkillIds().forEach(skillId -> skillService.assignSkillToUser(skillId, user.getId())));

vang31 replied:
если переменная не нужна, то зачем ею занимать память)

vang31 commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    public List<GoalDto> findSubtasksByGoalId(Long goalId, GoalFilterDto filters) {
+        Stream<Goal> subtasks = goalRepository.findByParent(goalId);
+
+        return goalFilters.stream()
```
так же стоит сразу в ретурне  ;

        return goalRepository.findByParent(goalId).stream()

sta1er replied:
Там логика другая немного, мы находим subtasks, чтобы потом их использовать по ходу стрима, который вызывается на goalFilters

vang31 replied:
не заметил, да теперь понял)

vang31 commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    public List<GoalDto> getGoalsByUser(Long userId, GoalFilterDto filters) {
+        Stream<Goal> goals = goalRepository.findGoalsByUserId(userId);
+
+        return goalFilters.stream()
+                .filter(filter -> filter.isApplicable(filters))
```
см комментарий сверху

sta1er replied:
Точно так же как и выше, мне кажется просто если всунуть "goalRepository.findGoalsByUserId(userId)" в середину, то будет не очень понятно потом что это и как это читать

sta1er replied:
Вообщем я посидел, подумал и понял что это скорее вопрос: "Не понизит ли это сильно уровень читаемость если я засуну "goalRepository.findGoalsByUserId(userId);" в середину другого стрима?"

vang31 replied:
это я не внимательно читал код)
читаемость будет ниже ты прав

vang31 commented on src/main/java/school/faang/user_service/service/skill/SkillService.java:
```diff
@@ -0,0 +1,44 @@
+    private SkillRepository skillRepository;
+
+    public boolean existsByTitle(String title) {
+        return skillRepository.existsByTitle(title);
+    }
+
+    public int countExisting(List<Long> ids) {
+        return skillRepository.countExisting(ids);
+    }
+
+    public List<Skill> findAllByUserId(long userId) {
+        return skillRepository.findAllByUserId(userId);
+    }
+
+    public List<Skill> findSkillsOfferedToUser(long userId) {
+        return skillRepository.findSkillsOfferedToUser(userId);
+    }
+
+    public Optional<Skill> findUserSkill(long skillId, long userId) {
+        return skillRepository.findUserSkill(skillId, userId);
+    }
+
+    public void assignSkillToUser(long skillId, long userId) {
+        skillRepository.assignSkillToUser(skillId, userId);
+    }
+
+    public List<Skill> findSkillsByGoalId(long goalId) {
+        return skillRepository.findSkillsByGoalId(goalId);
```
следует добавить валидацию, защита от дурачка. Код тем самым станет более безопасным и отказоустойчивее.

sta1er replied:
Это заглушка, я думаю ребята там уже у себя красиво сделают. Там же будет уже их логика и обработка ошибок

vang31 commented on src/test/java/school/faang/user_service/service/goal/FilterTest.java:
```diff
@@ -0,0 +1,186 @@
+import static org.mockito.Mockito.when;
+
+
+public class FilterTest {
```
стоит добавить негативную проверку на фильтры.

vang31 commented on src/test/java/school/faang/user_service/service/goal/GoalServiceTest.java:
```diff
@@ -0,0 +1,279 @@
+    }
+
+    @Test
+    void createGoal_Success() {
```
в java нет snakeCase) тольуо camelCase)

vang31 commented on src/test/java/school/faang/user_service/service/goal/GoalServiceTest.java:
```diff
@@ -0,0 +1,279 @@
+    }
+
+    @Test
+    void updateGoal_Success() {
```
updateGoalSuccessTest будет более звучащим)

vang31 commented on src/test/java/school/faang/user_service/service/goal/GoalServiceTest.java:
```diff
@@ -0,0 +1,279 @@
+    }
+
+    @Test
+    void updateGoal_ThrowsException_WhenGoalNotFound() {
```
можно также назвать goalUpdateWithExceptionWhenGoalNotFoundTest

vang31 reviewed (commented):
Поработай над названием тестов)

Commit:
внес правки
